### PR TITLE
Add partial support to the binary decoding initializer.

### PR DESCRIPTION
**Pull request by thomasvl**
Also add tests to confirm partial and full decoding work as
expected.

### CODE REVIEW

**Comment by thomasvl** [comment]
Added second commit with docs.

File: Sources/SwiftProtobuf/BinaryTypeAdditions.swift [hunk 1]
```diff
@@ -268,7 +268,20 @@ public extension Message {
         return visitor.serializedSize
     }
 
-    init(serializedData data: Data, extensions: ExtensionSet? = nil) throws {
+    /// Initializes the message by decoding the Protocol Buffer binary serialization
+    /// format into for this message.
```
Comment (allevato) [inline Sources/SwiftProtobuf/BinaryTypeAdditions.swift]: remove "into"

File: Sources/SwiftProtobuf/BinaryTypeAdditions.swift [hunk 2]
```diff
@@ -268,7 +268,20 @@ public extension Message {
...
+    ///
+    /// - Parameters:
+    ///   - serializedData: The binary serialation data to decode.
```
Comment (allevato) [inline Sources/SwiftProtobuf/BinaryTypeAdditions.swift]: s/serialation/serialization/

File: Sources/SwiftProtobuf/BinaryTypeAdditions.swift [hunk 3]
```diff
@@ -268,7 +268,20 @@ public extension Message {
...
+    ///   - extensions: An `ExtensionSet` to log up and decode any extensions in this
```
Comment (allevato) [inline Sources/SwiftProtobuf/BinaryTypeAdditions.swift]: s/log/look/?

File: Sources/SwiftProtobuf/BinaryTypeAdditions.swift [hunk 4]
```diff
@@ -268,7 +268,20 @@ public extension Message {
...
+    ///     message or message nested within this message's fields.
```
Comment (allevato) [inline Sources/SwiftProtobuf/BinaryTypeAdditions.swift]: "message or messages nested" (add "s")

File: Sources/SwiftProtobuf/BinaryTypeAdditions.swift [hunk 5]
```diff
@@ -268,7 +268,20 @@ public extension Message {
...
+    ///   - partial: The binary serialization format requires all `required` fields
```
Comment (allevato) [inline Sources/SwiftProtobuf/BinaryTypeAdditions.swift]: "By default, the binary serialization format..."

File: Sources/SwiftProtobuf/BinaryTypeAdditions.swift [hunk 6]
```diff
@@ -268,7 +268,20 @@ public extension Message {
...
+    ///     be present; when `partial` is `false`,
+    ///     `BinaryDecodingError.missingRequiredFields` is throw if any were missing.
```
Comment (allevato) [inline Sources/SwiftProtobuf/BinaryTypeAdditions.swift]: s/throw/thrown/

File: Sources/SwiftProtobuf/BinaryTypeAdditions.swift [hunk 7]
```diff
@@ -268,7 +268,20 @@ public extension Message {
...
+    ///     When `partial` is `true`, then partial messages are allowed, and
+    ///     `Message.isRequired` is not checked.
```
Comment (allevato) [inline Sources/SwiftProtobuf/BinaryTypeAdditions.swift]: s/isRequired/isInitialized/

File: Sources/SwiftProtobuf/BinaryTypeAdditions.swift [hunk 8]
```diff
@@ -268,7 +268,20 @@ public extension Message {
...
+    /// - Throws: An instance of `BinaryDecodingError` on failure .
```
Comment (allevato) [inline Sources/SwiftProtobuf/BinaryTypeAdditions.swift]: remove space before period

File: Sources/SwiftProtobuf/BinaryDecodingError.swift [hunk 9]
```diff
@@ -23,4 +23,8 @@ public enum BinaryDecodingError: Error {
     case malformedProtobuf
     /// The data being parsed does not match the type specified in the proto file
     case schemaMismatch
+    /// The message or sub messages definitions have required fields, and the
```
Comment (allevato) [inline Sources/SwiftProtobuf/BinaryDecodingError.swift]: s/sub/nested/

File: Sources/SwiftProtobuf/BinaryTypeAdditions.swift (hunk 1 above)
Comment (thomasvl) [inline Sources/SwiftProtobuf/BinaryTypeAdditions.swift]: Done.

File: Sources/SwiftProtobuf/BinaryTypeAdditions.swift (hunk 2 above)
Comment (thomasvl) [inline Sources/SwiftProtobuf/BinaryTypeAdditions.swift]: Done

File: Sources/SwiftProtobuf/BinaryTypeAdditions.swift (hunk 3 above)
Comment (thomasvl) [inline Sources/SwiftProtobuf/BinaryTypeAdditions.swift]: Yup, done.

File: Sources/SwiftProtobuf/BinaryTypeAdditions.swift (hunk 4 above)
Comment (thomasvl) [inline Sources/SwiftProtobuf/BinaryTypeAdditions.swift]: done

File: Sources/SwiftProtobuf/BinaryTypeAdditions.swift (hunk 5 above)
Comment (thomasvl) [inline Sources/SwiftProtobuf/BinaryTypeAdditions.swift]: Done.

File: Sources/SwiftProtobuf/BinaryTypeAdditions.swift (hunk 6 above)
Comment (thomasvl) [inline Sources/SwiftProtobuf/BinaryTypeAdditions.swift]: Done.

File: Sources/SwiftProtobuf/BinaryTypeAdditions.swift (hunk 7 above)
Comment (thomasvl) [inline Sources/SwiftProtobuf/BinaryTypeAdditions.swift]: Doh.

File: Sources/SwiftProtobuf/BinaryTypeAdditions.swift (hunk 8 above)
Comment (thomasvl) [inline Sources/SwiftProtobuf/BinaryTypeAdditions.swift]: Done.

File: Sources/SwiftProtobuf/BinaryDecodingError.swift (hunk 9 above)
Comment (thomasvl) [inline Sources/SwiftProtobuf/BinaryDecodingError.swift]: Done.

**Pull request merged**